blueprints/cloud-operations/uptime-monitoring-tool/source/net-mon-agent/main.py

```diff
@@ -0,0 +1,160 @@
+#!/usr/bin/env python3
+# Copyright 2023 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from google.api import label_pb2 as ga_label
+from google.api import metric_pb2 as ga_metric
+from google.api.metric_pb2 import MetricDescriptor
+from google.cloud import monitoring_v3
+from monitoring.utility import create_custom_metric, get_custom_metric
+
+import time
+import socket
+import click
+import logging
+import concurrent.futures
+
+# Set the log level
+logging.basicConfig(level=logging.INFO)
+LOGGER = logging.getLogger(__name__)
+
+client = monitoring_v3.MetricServiceClient()
+
+ENDPOINT_UPTIME_METRIC = "endpoint_uptime_check_passed"
+ENDPOINT_LATENCY = "endpoint_latency"
+
+
+class UptimeCheckResult:
+    """A class to represent the result of an uptime check."""
+
+    def __init__(self, host: str, port: int, latency: int, reachable: bool):
+        """Initializes a new UptimeCheckResult object.
+
+        Args:
+          host: The endpoint that was checked.
+          port: The port that was checked.
+          latency: The latency of the check in milliseconds.
+          reachable: Whether or not the endpoint was reachable.
+        """
+        self.host = host
+        self.port = port
+        self.latency = latency
+        self.reachable = reachable
+
+    def __str__(self):
+        """Returns a string representation of the UptimeCheckResult object."""
+        return f"UptimeCheckResult(endpoint={self.endpoint}, port={self.port}, latency={self.latency}, reachable={self.reachable})"
+
+
+def check_host_connectivity(timeout: int, host: str, port: int) -> UptimeCheckResult:
+    """Checks the connectivity to a given host and port.
+
+    Args:
+      timeout: Seconds to wait for the connection to be established
+      host: The hostname or IP address of the host to check.
+      port: The port number to check.
+
+    Returns:
+      An UptimeCheckResult object representing the result of the check.
+    """
+
+    start_time = time.perf_counter_ns()
+    try:
+        socket.create_connection((host, port), timeout).close()
+        latency = (time.perf_counter_ns() - start_time) // 1000000
+        return UptimeCheckResult(host=host, port=port, latency=latency, reachable=True)
```

Comment:
can you return using an f-string directly here? using a class seems overkill to do one formatting of three values...

Reply:
Updated the code, still return an uptime result which later computes one or more timeseries (according to the result) to write on Cloud Monitoring. lmkwyt!